lib/afinter.c

```diff
@@ -635,3 +643,107 @@ afinter_get_metrics(void)
 {
   return metrics;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  AFInterSource *self;
+
+  g_mutex_lock(&internal_msg_lock);
+
+  if (is_live_collection)
+    return AFINTER_LIVE_COLLECTION_RUNNING;
```

Comment:
@HofiOne : There is an unlock function call before the end of this function.

Reply:
but you might return from the function here, before reaching that unlock ever

Reply:
@HofiOne : okay, I have put lock after both `if` conditions.

Reply:
The purpose of locking here is to protect the internal objects from concurrent access (read/write), this way you leave `is_live_collection` and `current_internal_source` unprotected, and concurrent threads can modify them between entering into the `afinter_start_live_collection` and the `g_mutex_lock(&internal_msg_lock)` can take effect in line 661.

e.g. this can be a better approach

```
AFInterLive
afinter_start_live_collection(void)
{
  g_mutex_lock(&internal_msg_lock);

  if (is_live_collection)
    {
      g_mutex_unlock(&internal_msg_lock);
      return AFINTER_LIVE_COLLECTION_RUNNING;
    }
  if (current_internal_source != NULL)
    {
      g_mutex_unlock(&internal_msg_lock);
      return AFINTER_INTERNAL_SRC_PRESENT;
    }

  AFInterSource *self = g_new0(AFInterSource, 1);
  AFInterSourceOptions *options = g_new0(AFInterSourceOptions, 1);

  afinter_source_options_defaults(options);
  self->options = options;
  stats_counter_set(metrics.queue_capacity, self->options->queue_capacity);

  current_internal_source = self;
  is_live_collection = TRUE;

  g_mutex_unlock(&internal_msg_lock);

  return AFINTER_LIVE_COLLECTION_STARTED;
}
```

Reply:
@HofiOne : okay, thanks for the code snipped.